src/analysis/calculus/inverse.lean

```diff
@@ -0,0 +1,582 @@
+((hf.implicit_to_local_homeomorph f f' f'inv hf').eventually_right_inverse $
+  hf.mem_implicit_to_local_homeomorph_target hf').mono $ λ ⟨z, y⟩ h,
+    congr_arg prod.fst h
+
+/-- Any point in some neighborhood of `a` can be represented as `implicit_function`
+of some point. -/
+lemma eq_implicit_function :
+  ∀ᶠ x in 𝓝 a, hf.implicit_function f f' f'inv hf' (f x)
+    ((hf.implicit_to_local_homeomorph f f' f'inv hf').to_fun x).snd = x :=
+(hf.implicit_aux_has_fderiv hf').eventually_left_inverse
+
+/-- Derivative of the inverse function used to prove the implicit function theorem. -/
+lemma to_implicit_function_aux :
+  has_strict_fderiv_at (hf.implicit_to_local_homeomorph f f' f'inv hf').inv_fun
+    (f'inv.coprod $ subtype_val f'.ker) (f a, 0) :=
+hf.implicit_to_local_homeomorph_self hf' ▸
+  (hf.implicit_aux_has_fderiv hf').to_local_inverse
+
+lemma to_implicit_function :
+  has_strict_fderiv_at (hf.implicit_function f f' f'inv hf' (f a)) (subtype_val f'.ker) 0 :=
+begin
+  have := (hf.to_implicit_function_aux hf').comp 0
+    ((has_strict_fderiv_at_const (f a) 0).prod $ has_strict_fderiv_at_id 0),
+  convert this,
+  ext x,
+  simp
+end
+
+end has_strict_fderiv_at
+
+namespace has_strict_fderiv_at
+
+variables [cs : complete_space (E × F)] {f : E × F → G} (f' : E × F →L[𝕜] G) (f'inv : G →L[𝕜] F)
+  {p : E × F} (hf : has_strict_fderiv_at f f' p)
+  (hf'l : ∀ y : F, f'inv (f' (0, y)) = y) (hf'r : ∀ z : G, f' (0, f'inv z) = z)
+
+/-- Formula for the derivative of an implicit function. -/
+def prod_implicit_function_fderiv : (E × G) →L[𝕜] F :=
+(f'inv.comp $ continuous_linear_map.snd 𝕜 E G -
+      f'.comp ((continuous_linear_map.id 𝕜 E).prod_map 0))
+
+variables {f' f'inv}
+
+@[simp] lemma prod_implicit_function_fderiv_apply (x) :
+  prod_implicit_function_fderiv f' f'inv x = f'inv (x.2 - f' (x.1, 0)) := rfl
+
+include hf'r
+
+lemma prod_implicit_fderiv_right_inverse (x : E) (z : G) :
+  f' (x, f'inv z) = f' (x, 0) + z :=
+by { conv_rhs { rw [← hf'r z] }, simp [← f'.map_add] }
+
+include hf'l
+
+variables (f' f'inv)
+
+/-- Derivative of an auxiliary function used in the proof of the implicit function theorem. -/
+def prod_implicit_function_aux_fderiv : (E × F) ≃L[𝕜] (E × G) :=
+continuous_linear_equiv.equiv_of_inverse
+  ((continuous_linear_map.fst 𝕜 E F).prod f')
+  ((continuous_linear_map.fst 𝕜 E G).prod $ prod_implicit_function_fderiv f' f'inv)
+  (λ ⟨x, y⟩, by simp [← continuous_linear_map.map_sub, hf'l])
+  (λ ⟨x, z⟩, by simp [-continuous_linear_map.map_sub, prod_implicit_fderiv_right_inverse hf'r])
+
+variables {f' f'inv}
+
+include hf
+
+lemma prod_implicit_function_aux_deriv :
+  has_strict_fderiv_at (λ x : E × F, (x.1, f x))
+    (prod_implicit_function_aux_fderiv f' f'inv hf'l hf'r : (E × F) →L[𝕜] E × G) p :=
+has_strict_fderiv_at_fst.prod hf
+
+include cs
+variable (f)
+
+/-- Implicit function `g` defined by an equation `f (x, g(x, y)) = z`. -/
+def prod_implicit_function (x : E × G) : F :=
```

Comment:
Could you restate the assumptions in the definition (i.e., the prop bits in the section variables)? Otherwise, the assumptions for the proposition are hard to decipher.